src/safemode_ui.h

```diff
@@ -0,0 +1,119 @@
+#ifndef SEFEMODE_UI_H
+#define SEFEMODE_UI_H
+
+#include <unordered_map>
+#include <string>
+#include <vector>
+#include <locale>
+#include <algorithm>
+#include "json.h"
+#include "enums.h"
+#include "creature.h"
+
+class safemode : public JsonSerializer, public JsonDeserializer
+{
+    private:
+        void test_pattern( const int iCurrentPage, const int iCurrentLine );
+
+        void load( const bool bCharacter );
+        bool save( const bool bCharacter );
+
+        bool bChar;
+
+        enum TAB : int {
+            GLOBAL_TAB,
+            CHARACTER_TAB,
+            MAX_TAB
+        };
+
+        class cRules
+        {
+            public:
+                std::string sRule;
+                bool bActive;
+                bool bExclude;
+                int attCreature;
+                int iProxyDist;
+
+                cRules() {
+                    this->sRule = "";
+                    this->bActive = false;
+                    this->bExclude = false;
+                    this->attCreature = Creature::A_HOSTILE;
+                    this->iProxyDist = 0;
+                }
+
+                cRules( std::string sRuleIn, bool bActiveIn, bool bExcludeIn, int attCreatureIn,
+                        int iProxyDistIn ) {
+                    this->sRule = sRuleIn;
+                    this->bActive = bActiveIn;
+                    this->bExclude = bExcludeIn;
+                    this->attCreature = attCreatureIn;
+                    this->iProxyDist = iProxyDistIn;
+                }
+
+                ~cRules() {};
+        };
+
+        class cRuleState
+        {
+            public:
+                rule_state state;
+                int proxy_dist;
+
+                cRuleState() {
+                    this->state = RULE_NONE;
+                    this->proxy_dist = 0;
+                }
+
+                cRuleState( rule_state state_in, int proxy_dist_in ) {
+                    this->state = state_in;
+                    this->proxy_dist = proxy_dist_in;
+                }
+        };
+
+        /**
+         * The currently-active set of safemode rules, in a form that allows quick
+         * lookup. When this is filled (by @ref safemode::create_rules()), every
+         * monster existing in the game that matches a rule (either white- or blacklist)
+         * is added as the key, with RULE_WHITELISTED or RULE_BLACKLISTED as the values.
+         * map_monsters[ 'mob name' ][ 'monAttitude' ].cRuleState('rule_state', 'proxy dist')
+         */
+        std::unordered_map < std::string, std::array < cRuleState, Creature::A_MAX - 1 > > map_monsters;
+
+        /**
+         * - vRules[0,1] aka vRules[GLOBAL,CHARACTER]: current rules split into global and
+         *      character-specific. Allows the editor to show one or the other.
+         */
+        std::array<std::vector<cRules>, MAX_TAB> vRules;
```

Comment:
Using arrays with enumed indices is hard to maintain. It also makes bad practice seem better by not requiring extracting functions.

Consider splitting them up into character rules and global rules.